### DISCUSSION TITLE
Primeng 18 with Custom Preset Issues (always dark mode)

### DISCUSSION THREAD

**Discussion by pipodesignstudio** [primefaces/community#2860] [PrimeNG]
I'm running into issues customizing the PrimeNG 18 theme. I'm trying to replicate the grayscale reference theme from the documentation but something's not quite working. I've started by configuring the theme initialization in app.component.ts. Like so: 

```
import { Component } from '@angular/core';
import { RouterOutlet } from '@angular/router';
import { PrimeNGConfig } from 'primeng/api';
import { CustomTheme } from './common/theme/custom-theme.preset';

@Component({
  selector: 'app-root',
  standalone: true,
  imports: [RouterOutlet],
  template: ` <router-outlet /> `,
})
export class AppComponent {
  title = 'demo-site';
  constructor(private primengConfig: PrimeNGConfig) {
    this.primengConfig.theme.set({
      preset: CustomTheme,
      options: {
        cssLayer: {
          name: 'primeng',
          order: 'tailwind-base, primeng, tailwind-utilities'
        },
      },
    });
  }
}
```

Custom Theme looks like this (code extracted from official docs):

```
import { definePreset } from 'primeng/themes';
import { Aura } from 'primeng/themes/aura';

export const CustomTheme = definePreset(Aura, {
    semantic: {
        primary: {
            50: '{indigo.50}',
            100: '{indigo.100}',
            200: '{indigo.200}',
            300: '{indigo.300}',
            400: '{indigo.400}',
            500: '{indigo.500}',
            600: '{indigo.600}',
            700: '{indigo.700}',
            800: '{indigo.800}',
            900: '{indigo.900}',
            950: '{indigo.950}'
        }
    },
    colorScheme: {
        light: {
            primary: {
                color: '{zinc.950}',
                inverseColor: '#ffffff',
                hoverColor: '{zinc.900}',
                activeColor: '{zinc.800}'
            },
            highlight: {
                background: '{zinc.950}',
                focusBackground: '{zinc.700}',
                color: '#ffffff',
                focusColor: '#ffffff'
            }
        },
        dark: {
            primary: {
                color: '{zinc.50}',
                inverseColor: '{zinc.950}',
                hoverColor: '{zinc.100}',
                activeColor: '{zinc.200}'
            },
            highlight: {
                background: 'rgba(250, 250, 250, .16)',
                focusBackground: 'rgba(250, 250, 250, .24)',
                color: 'rgba(255,255,255,.87)',
                focusColor: 'rgba(255,255,255,.87)'
            }
        }
    }
})
```

There are two key files in my projects:

```
/* styles.scss */
@layer tailwind-base, primeng, tailwind-utilities;

@layer tailwind-base {
@tailwind base;
}

@layer tailwind-utilities {
@tailwind components;
@tailwind utilities;
}

* {
    @apply font-poppins;
}

```
and: 
```
/** @type {import('tailwindcss').Config} */
module.exports = {
  content: ["./src/**/*.{html,ts}"],
  darkMode: 'false',
  theme: {
    extend: {
      fontFamily: {
        poppins: ["Poppins"]
      }
    },
  },
  plugins: [require('tailwindcss-primeui')]

}
```

I'm experiencing difficulties applying custom styles using the PrimeNG theme. While some preset styles are working as expected, most of my customizations, whether at a global level or within individual components, are being ignored. Despite not selecting the dark theme, some dark theme styles are being applied. The light theme seems to be inactive. There's no overarching class controlling the theme. Additionally, my custom CSS, especially from @reset, is being overridden by the theme's styles. As a temporary solution, I'm using Tailwind's styleClass on individual components, but this is not a scalable approach.

**Comment by aboudard** [comment]
Hello ! 
I couldn't anwser your question for now, but I have one for you, did you manage to make a dark/light theme switch with primeng 18 ?
I know you say that you have issues, but is there some case / component that does react to a top level class switch like in the documentation pages ? `p-dark` for example.
Thanks !
Alain

**Reply by pipodesignstudio** [reply]
Hi @aboudard thanks for your reply. After diving into this yesterday, including watching a Vue tutorial (https://www.youtube.com/watch?v=J3KFw5sih98&t=1140s) and exploring Aura's types, I've figured out what's going on. It seems my browser's prefers-color-scheme setting is forcing dark mode, even though I'm not explicitly using Tailwind classes or a dark mode service. Consequently, only the dark mode customizations in my preset configuration are taking effect.

This presents some technical challenges. I attempted to bypass the colorScheme property and set a global style for the component, expecting it to work regardless of the theme. However, this approach has been unsuccessful.

```
/* Won't work */
components: {
button: {
      colorScheme: {
        dark: {
          primary: {
            background: '{primary.900}',
            color: '{primary.50}',
            hoverBackground: '{primary.950}',
            hoverColor: '{surface.50}',
          },
        },
      },
    }
}
```
In turn, this will achieve the desired outcome.

```
components: {
button: {
      colorScheme: {
        dark: {
          primary: {
            background: '{primary.900}',
            color: '{primary.50}',
            hoverBackground: '{primary.950}',
            hoverColor: '{surface.50}',
          },
        },
      }
    }
}
```
PrimeNG's theming options offer a lot of flexibility. But the lack of a global theme without mode switching means I have to duplicate the scheme for dark and light modes. I'm also facing challenges with applying styles at higher levels, like the semantic formField. I'll keep looking into it.

**Reply by aboudard** [reply]
Hi @pipodesignstudio !
This is interesting, on my side, I managed to make the initial switch as described in the doc.

Since the configuration object is so big, it's easy to make a mistake, and there is no validation of this object, this is unfortunate.
I hope the final version will provide a type validation of the configuration.

Cheers

**Reply by qwadrox** [reply]
> 's going on. It seems my browser's prefers-color-scheme setting is forcing dark mode, even though I'm not explicitly using Tailwind classes or a dark

PrimeNG uses the browser setting as the default darkModeSelector in theme configuration
https://v18.primeng.org/theming#darkmode
You can control which mode, dark or light, takes place.

**Reply by aboudard** [reply]
> PrimeNG uses the browser setting as the default darkModeSelector in theme configuration https://v18.primeng.org/theming#darkmode You can control which mode, dark or light, takes place.

Well, I managed to swith dark / light with a custom class applied on root `html` tag, but with Chrome settings only, I did not succeed. Maybe I missed something.

**Reply by qwadrox** [reply]
@aboudard 
https://stackblitz.com/edit/badqp6?file=src%2Fapp%2Fbutton-severity-demo.html,src%2Fapp%2Fbutton-severity-demo.ts
I have this example with Primeng 18 and I can switch between dark and light with Chrome Devtools. (If I download the code from stackblitz it works also by changing the the preferences in the browser.) 
<img width="1079" alt="image" src="https://github.com/user-attachments/assets/e686719b-aac4-490b-9506-b6dc2e4a83e6">

**Reply by aboudard** [reply]
Thanks @qwadrox ! I'll try that as a complement of my class mode !

**Comment by qwadrox** [comment]
> , {
>     semantic: {
>         primary: {
>             50: '{indigo.50}',
>             100: '{indigo.100}',
>             200: '{indigo.200}',
>             300: '{indigo.300}',
>             400: '{indigo.400}',
>             500: '{indigo.500}',
>             600: '{indigo.600}',
>             700: '{indigo.700}',
>             800: '{indigo.800}',
>             900: '{indigo.900}',
>             950: '{indigo.950}'
>         }
>     },
>     colorScheme: {
>         light: {
>             primary: {
>                 color: '{zinc.950}',
>                 inverseColor: '#ffffff',
>                 hoverColor: '{zinc.900}',
>                 activeColor: '{zinc.800}'
>             },
>             highlight: {
>                 background: '{zinc.950}',
>                 focusBackground: '{zinc.700}',
>                 color: '#ffffff',
>                 focusColor: '#ffffff'
>             }
>         },
>         dark: {
>             primary: {
>                 color: '{zinc.50}',
>                 inverseColor: '{zinc.950}',
>                 hoverColor: '{zinc.100}',
>                 activeColor: '{zinc.200}'
>             },
>             highlight: {
>                 background: 'rgba(250, 250, 250, .16)',
>                 focusBackground: 'rgba(250, 250, 250, .24)',
>                 color: 'rgba(255,255,255,.87)',
>                 focusColor: 'rgba(255,255,255,.87)'
>             }
>         }
>     }

**Always try to console.log the preset you are editing**. The provided example in the docs not working because the `colorScheme` is misplaced. It should be `semantic.colorScheme`; 
Look: https://stackblitz.com/edit/badqp6?file=src%2Fapp%2Fbutton-severity-demo.ts

This version is still beta, so both the docs and components might have issues.
